File: cairnkit/stages.py
"""The full 16-stage delivery pipeline metadata (M2, completed in B3).

Single source of truth for: the canonical stage order, each stage's produced artifact,
which stages are CLARIFY pauses, which are retryable verify gates, and the path-mode
stage sets (IntentGate routes small work onto a shorter path). state.py/gate.py consume this.
"""

from __future__ import annotations

# Canonical order. advance() walks this, skipping stages not in the active path mode.
FULL_SEQUENCE = (
    "INIT",
    "INTENT_GATE",
    "ANALYSE_PRODUCT",
    "CLARIFY_PRODUCT",
    "ANALYSE_TECH",
    "CLARIFY_TECH",
    "ARCHITECT_BACKEND",
    "CLARIFY_ARCH_BACKEND",
    "ARCHITECT_FRONTEND",
    "CLARIFY_ARCH_FRONTEND",
    "IMPLEMENT",
    "BUILD_VERIFY",
    "VISUAL_REVIEW",
    "E2E_VERIFY",
    "TEST",
    "ARCHIVE",
    "DONE",
)

# Artifact a stage produces (verified on advance out of that stage). Others produce nothing.
STAGE_ARTIFACT = {
    "ANALYSE_PRODUCT": "01-product.md",
    "ANALYSE_TECH": "02-tech.md",
    "ARCHITECT_BACKEND": "03-arch.md",
    "ARCHITECT_FRONTEND": "04-arch-fe.md",
    "IMPLEMENT": "05-implement.md",
    "BUILD_VERIFY": "06-build.md",
    "VISUAL_REVIEW": "07-visual.md",
    "E2E_VERIFY": "08-e2e.md",
    "TEST": "09-test.md",
    "ARCHIVE": "10-archive.md",
}

CLARIFY_STAGES = frozenset({
    "CLARIFY_PRODUCT", "CLARIFY_TECH", "CLARIFY_ARCH_BACKEND", "CLARIFY_ARCH_FRONTEND",
})

# Verify gates that loop on failure with a retry cap before being blocked.
RETRY_STAGES = frozenset({"BUILD_VERIFY", "E2E_VERIFY"})
RETRY_CAP = 5

# The role agent the orchestrator dispatches at each stage (Markdown layer).
STAGE_AGENT = {
    "ANALYSE_PRODUCT": "product",
    "ANALYSE_TECH": "tech",
    "ARCHITECT_BACKEND": "architect-be",
    "ARCHITECT_FRONTEND": "architect-fe",
    "IMPLEMENT": "dev",
    "BUILD_VERIFY": "verify",
    "VISUAL_REVIEW": "visual",
    "E2E_VERIFY": "verify",
    "TEST": "verify",
    "ARCHIVE": "archiver",
}

PATH_MODES = ("full", "lite", "single")

# Stages excluded from the shorter paths (IntentGate routing).
_LITE_EXCLUDE = frozenset({"ARCHITECT_FRONTEND", "CLARIFY_ARCH_FRONTEND", "VISUAL_REVIEW"})
_SINGLE_INCLUDE = frozenset({
    "INIT", "INTENT_GATE", "IMPLEMENT", "BUILD_VERIFY", "TEST", "ARCHIVE", "DONE",
})


def stages_for(path_mode: str) -> tuple[str, ...]:
    """The ordered stages active for a path mode. Raises on an unknown mode (no silent fallback)."""
    if path_mode == "single":
        return tuple(s for s in FULL_SEQUENCE if s in _SINGLE_INCLUDE)
    if path_mode == "lite":
        return tuple(s for s in FULL_SEQUENCE if s not in _LITE_EXCLUDE)
    if path_mode == "full":
        return FULL_SEQUENCE
    raise ValueError(f"unknown path_mode {path_mode!r}; valid: {PATH_MODES}")


def next_stage(current: str, path_mode: str) -> str | None:
    """The next active stage after ``current`` for the path mode, or None at the end."""
    active = stages_for(path_mode)
    if current not in active:
        # current was skipped/foreign — fall back to its position in the full sequence
        idx = FULL_SEQUENCE.index(current)
        for s in FULL_SEQUENCE[idx + 1:]:
            if s in active:
                return s
        return None
    i = active.index(current)
    return active[i + 1] if i + 1 < len(active) else None


def is_clarify(stage: str) -> bool:
    return stage in CLARIFY_STAGES

File: src/stages.rs
//! The full 16-stage delivery pipeline metadata (mirrors Python `cairnkit.stages`).

/// Canonical stage order. advance() walks this, skipping stages not in the active path mode.
pub const FULL_SEQUENCE: &[&str] = &[
    "INIT",
    "INTENT_GATE",
    "ANALYSE_PRODUCT",
    "CLARIFY_PRODUCT",
    "ANALYSE_TECH",
    "CLARIFY_TECH",
    "ARCHITECT_BACKEND",
    "CLARIFY_ARCH_BACKEND",
    "ARCHITECT_FRONTEND",
    "CLARIFY_ARCH_FRONTEND",
    "IMPLEMENT",
    "BUILD_VERIFY",
    "VISUAL_REVIEW",
    "E2E_VERIFY",
    "TEST",
    "ARCHIVE",
    "DONE",
];

pub const PATH_MODES: &[&str] = &["full", "lite", "single"];
pub const RETRY_STAGES: &[&str] = &["BUILD_VERIFY", "E2E_VERIFY"];
pub const RETRY_CAP: u32 = 5;

const LITE_EXCLUDE: &[&str] = &["ARCHITECT_FRONTEND", "CLARIFY_ARCH_FRONTEND", "VISUAL_REVIEW"];
const SINGLE_INCLUDE: &[&str] = &[
    "INIT", "INTENT_GATE", "IMPLEMENT", "BUILD_VERIFY", "TEST", "ARCHIVE", "DONE",
];

/// Artifact a stage produces (verified on advance out of that stage). None for others.
pub fn stage_artifact(stage: &str) -> Option<&'static str> {
    match stage {
        "ANALYSE_PRODUCT" => Some("01-product.md"),
        "ANALYSE_TECH" => Some("02-tech.md"),
        "ARCHITECT_BACKEND" => Some("03-arch.md"),
        "ARCHITECT_FRONTEND" => Some("04-arch-fe.md"),
        "IMPLEMENT" => Some("05-implement.md"),
        "BUILD_VERIFY" => Some("06-build.md"),
        "VISUAL_REVIEW" => Some("07-visual.md"),
        "E2E_VERIFY" => Some("08-e2e.md"),
        "TEST" => Some("09-test.md"),
        "ARCHIVE" => Some("10-archive.md"),
        _ => None,
    }
}

pub fn is_clarify(stage: &str) -> bool {
    stage.starts_with("CLARIFY")
}

pub fn is_valid_stage(stage: &str) -> bool {
    FULL_SEQUENCE.contains(&stage)
}

pub fn is_valid_path_mode(mode: &str) -> bool {
    PATH_MODES.contains(&mode)
}

/// The ordered stages active for a path mode. Returns None for an unknown mode (no silent fallback).
pub fn stages_for(path_mode: &str) -> Option<Vec<&'static str>> {
    match path_mode {
        "single" => Some(
            FULL_SEQUENCE
                .iter()
                .copied()
                .filter(|s| SINGLE_INCLUDE.contains(s))
                .collect(),
        ),
        "lite" => Some(
            FULL_SEQUENCE
                .iter()
                .copied()
                .filter(|s| !LITE_EXCLUDE.contains(s))
                .collect(),
        ),
        "full" => Some(FULL_SEQUENCE.to_vec()),
        _ => None,
    }
}

/// The next active stage after `current` for the path mode, or None at the end.
pub fn next_stage(current: &str, path_mode: &str) -> Option<&'static str> {
    let active = stages_for(path_mode)?;
    if let Some(i) = active.iter().position(|s| *s == current) {
        return active.get(i + 1).copied();
    }
    // current was skipped/foreign — fall back to its position in the full sequence
    let idx = FULL_SEQUENCE.iter().position(|s| *s == current)?;
    FULL_SEQUENCE[idx + 1..]
        .iter()
        .find(|s| active.contains(s))
        .copied()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn full_chain() {
        assert_eq!(next_stage("INIT", "full"), Some("INTENT_GATE"));
        assert_eq!(next_stage("ARCHIVE", "full"), Some("DONE"));
        assert_eq!(next_stage("DONE", "full"), None);
    }

    #[test]
    fn lite_skips_frontend() {
        assert_eq!(next_stage("CLARIFY_ARCH_BACKEND", "lite"), Some("IMPLEMENT"));
    }

    #[test]
    fn foreign_current_falls_back() {
        assert_eq!(next_stage("ARCHITECT_FRONTEND", "lite"), Some("IMPLEMENT"));
        assert_eq!(next_stage("CLARIFY_TECH", "single"), Some("IMPLEMENT"));
    }

    #[test]
    fn single_is_minimal() {
        let s = stages_for("single").unwrap();
        assert!(!s.contains(&"ANALYSE_PRODUCT"));
        assert!(s.contains(&"IMPLEMENT") && s.contains(&"DONE"));
    }

    #[test]
    fn unknown_mode_none() {
        assert!(stages_for("turbo").is_none());
    }
}
